torch_sim/optimizers.py

```diff
@@ -445,6 +445,252 @@ def gd_step(
+        atom_wise_dt = state.dt[state.batch].unsqueeze(-1)
+        state.velocities += 0.5 * atom_wise_dt * state.forces / state.masses.unsqueeze(-1)
+
+        # Split positions and forces into atomic and cell components
+        atomic_positions = state.positions  # shape: (n_atoms, 3)
+
+        # Update atomic positions
+        atomic_positions_new = atomic_positions + atom_wise_dt * state.velocities
+
+        # Update state with new positions and cell
+        state.positions = atomic_positions_new
+
+        # Get new forces, energy, and stress
+        results = model(state)
+        state.energy = results["energy"]
+        state.forces = results["forces"]
+
+        # Velocity Verlet first half step (v += 0.5*a*dt)
+        state.velocities += 0.5 * atom_wise_dt * state.forces / state.masses.unsqueeze(-1)
+
+        # Calculate power (F·V) for atoms
+        atomic_power = (state.forces * state.velocities).sum(dim=1)  # [n_atoms]
+        atomic_power_per_batch = torch.zeros(
+            n_batches, device=device, dtype=atomic_power.dtype
+        )
+        atomic_power_per_batch.scatter_add_(
+            dim=0, index=state.batch, src=atomic_power
+        )  # [n_batches]
+
+        # Calculate power for cell DOFs
+        batch_power = atomic_power_per_batch
+
+        for batch_idx in range(n_batches):
+            # FIRE specific updates
+            if batch_power[batch_idx] > 0:  # Power is positive
+                state.n_pos[batch_idx] += 1
+                if state.n_pos[batch_idx] > n_min:
+                    state.dt[batch_idx] = min(state.dt[batch_idx] * f_inc, dt_max)
+                    state.alpha[batch_idx] = state.alpha[batch_idx] * f_alpha
+            else:  # Power is negative
+                state.n_pos[batch_idx] = 0
+                state.dt[batch_idx] = state.dt[batch_idx] * f_dec
+                state.alpha[batch_idx] = alpha_start[batch_idx]
+                # Reset velocities for both atoms and cell
+                state.velocities[state.batch == batch_idx] = 0
+
+        # Mix velocity and force direction using FIRE for atoms
+        v_norm = torch.norm(state.velocities, dim=1, keepdim=True)
+        f_norm = torch.norm(state.forces, dim=1, keepdim=True)
+        # Avoid division by zero
+        # mask = f_norm > 1e-10
+        # state.velocity = torch.where(
+        #     mask,
+        #     (1.0 - state.alpha) * state.velocity
+        #     + state.alpha * state.forces * v_norm / f_norm,
+        #     state.velocity,
+        # )
+        batch_wise_alpha = state.alpha[state.batch].unsqueeze(-1)
+        state.velocities = (
+            1.0 - batch_wise_alpha
+        ) * state.velocities + batch_wise_alpha * state.forces * v_norm / (f_norm + eps)
+
+        return state
+
+    return fire_init, fire_step
+
+
 @dataclass
 class BatchedUnitCellFireState(SimState, DeformGradMixin):
     """State information for batched FIRE optimization with unit cell degrees of
```

Comment:
We could take this opportunity to rename `BatchedUnitCellFire` -> `UnitCellFire`, and also remove the `Batched` descriptor from the other SimStates in this file, since batching is the default.

We could also make this inherit from `FireState` instead of `SimState`.